bin/test.sh

```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+[ $# -eq 1 ] || { echo "Usage: $0 STACK"; exit 1; }
+
+STACK="${1}"
+
+# Converts eg "heroku-18" -> "heroku/heroku:18".
+RUNTIME_IMAGE="heroku/${STACK/-/:}"
+
+if [[ "${STACK}" == "cedar-14" ]]; then
+    BUILD_IMAGE="${RUNTIME_IMAGE}"
+else
+    BUILD_IMAGE="${RUNTIME_IMAGE}-build"
+fi
+
+OUTPUT_IMAGE="redis-test-${STACK}"
+
+echo "Building buildpack on stack ${STACK}..."
+
+docker build \
+    --build-arg "BUILD_IMAGE=${BUILD_IMAGE}" \
+    --build-arg "RUNTIME_IMAGE=${RUNTIME_IMAGE}" \
+    -t "${OUTPUT_IMAGE}" \
+    .
+
+echo "Checking redis-server responds to ping command..."
+
+# Redis <4 does not support connection URLs so REDIS_URL has to be parsed:
+# https://stackoverflow.com/questions/38271281/can-i-use-redis-cli-with-a-connection-url
+REDIS_CONNECTION_ARGS="echo \${REDIS_URL} | sed 's_redis://\(.*\):\(.*\)@\(.*\):\(.*\)_-h \3 -p \4 -a \2_'"
+TEST_COMMAND="source .profile.d/redis.sh && redis-cli \$(${REDIS_CONNECTION_ARGS}) ping"
+docker run --rm -it "${OUTPUT_IMAGE}" bash -c "${TEST_COMMAND}"
```

Comment:
Just to double-check, are you confident the correct exit code ends up coming out here? Just thinking:

* the `docker run` needs to return the exit code from the `bash -c ...` bit.
* the `bash -c` bit needs to return the exit code of `TEST_COMMAND`
* `redis-cli` needs to return a non-zero exit code if the connection fails.
* (possibly missed some steps, you see what I mean though)

I have to ask just because obviously it's quite important for this script to work as a CI thing!

Reply:
Testing this locally, the exit code works for everything but redis-cli, which led me to:
https://github.com/antirez/redis/issues/2270#issuecomment-76777000

I'm hopeful Redis 4/5 has fixed this, in which case we'll have coverage from those parts of the test matrix. If they haven't, I'll update this to do a grep for PONG instead.

Reply:
The exit code works fine in Redis 4 + 5 so going to leave this as-is for now (with the next PR we'll be testing against multiple versions, and REDIS_URL is set the same for all of the, so some coverage is fine). Hopefully we'll drop support for Redis 3 at some point in the future too.

Reply:
I don't quite follow. So if the Redis version is 3 then the script may fail silently? That doesn't sound very good.